scipy/interpolate/src/_dierckxmodule.cc

```diff
@@ -0,0 +1,775 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include <iostream>
+#include <string>
+#include <memory>
+#include "numpy/arrayobject.h"
+#include "__fitpack.h"
+
+
+/*
+ * Check obj to be an ndim C contiguous array of specified typenum
+ *
+ *  sort of mimic checks cython does for `double[::1] obj`
+ */
+static int
+check_array(PyObject *obj, npy_intp ndim, int typenum) {
+
+    int cond = (PyArray_CheckExact(obj) &&
+               (PyArray_TYPE((PyArrayObject*)obj) == typenum) &&
+               (PyArray_NDIM((PyArrayObject*)obj) == ndim) &&
+                PyArray_CHKFLAGS((PyArrayObject*)obj, NPY_ARRAY_ALIGNED | NPY_ARRAY_C_CONTIGUOUS)
+    );
+
+    if(!cond) {
+        std::string msg = ("Expected a " + std::to_string(ndim) + "-dim C contiguous array " +
+                           " of dtype = " + std::to_string(typenum) + "( got " +
+                           std::to_string(PyArray_TYPE((PyArrayObject*)obj)) +" )\n");
+        // XXX: name the dtype from typenum? Also type of arg if not array
+        PyErr_SetString(PyExc_ValueError, msg.c_str());
+        return 0;
+    }
+    return 1;
+}
+
+
+/*
+ * RAII malloc-d work arrays
+ */
+struct malloc_deleter {
+    void operator()(void *p) const { std::free(p); }
+};
+
+template <typename T>
+using unique_fptr = std::unique_ptr<T, malloc_deleter>;
```

Comment:
You're using this to wrap "c-style" arrays. I think it would be preferable to use ``std::array``/``std::vector`` instead. This will remove the need for this template entirely, as ``std::unique_ptr`` calls ``T``'s destructor by default.

Reply:
Makes sense! 
Pushed a commit where `wrk` arrays are allocated as `vector<double> wrk; wrk.resize(2*k+2)` instead of this, admittedly, fairly confusing incantation.